asapdiscovery-ml/asapdiscovery/ml/trainer.py

```diff
@@ -994,12 +1019,32 @@ def train(self):
                 for sp, sp_d in self.loss_dict.items()
             }
 
+        final_model_path = self.output_dir / "final.th"
+        torch.save(self.model.state_dict(), final_model_path)
+        (self.output_dir / "loss_dict.json").write_text(json.dumps(self.loss_dict))
+
+        if self.upload_to_s3:
+            print("Uploading to S3")
+            s3 = S3.from_settings(self.s3_settings)
+            s3.push_file(
+                final_model_path,
+                location=self.s3_path,
+                content_type="application/octet-stream",
+            )
+            if self.use_wandb:
+                # track S3 artifacts
+                print("Linking S3 artifacts to W&B")
+                model_artifact = wandb.Artifact(
```

Comment:
maybe I'm missing something, but does this encode any S3 information anywhere/do we want that to be the case? seems like we're just separately pushing to S3 and to W&B

Reply:
Ah hang on I think I am not doing the linking properly, ill revisit.

Reply:
Good spot.

Reply:
Should be better now.

Reply:
yeah looks good, thanks!